# Connecting MySQL Database to Jupyter Notebooks

### Step 1. Download mysql-connector-python 
You can do this through commands such as: pip install mysql-connector-python 

### Step 2. Connect to your database
Add your password and uncomment out the others

In [1]:
import pandas as pd
import mysql.connector

db = mysql.connector.connect(
    user="root",
    password="{Password}", #angy, comment/un-comment for yours
    #password = "ilo5517WD", 
    host="localhost",
    port = 3306,
    database = 'rideshare'
)

#### Checking connection

In [2]:
print(db)


#### Setting cursor that will allow us to execute commands like in MySQL queries

In [3]:
cursor = db.cursor() 

### Step 3. Import Tables 

### Listing all tables in current database = rideshare

In [4]:
cursor.execute("SHOW TABLES")
for table in [tables[0] for tables in cursor.fetchall()]:
    print(table)

april_green
april_yellow
feb_green_one_rider
feb_green_two_rider
february_green
february_yellow
jan_green_one_rider
jan_green_two_rider
january_green
january_yellow
june_green
june_yellow
mar_green_one_rider
mar_green_two_rider
march_green
march_yellow
may_green
may_yellow


### Selecting the january green taxi data that was split with passenger count = 1

In [5]:
jan_green_one_rider = pd.read_sql("select * from jan_green_one_rider", db);

In [6]:
jan_green_one_rider.head()

ride_id     pickup_datetime    dropoff_datetime  pickup_longitude  \
0       14 2014-01-01 00:00:07 2014-01-01 00:08:28        -73.916992   
1       16 2014-01-01 00:00:13 2014-01-01 00:07:26        -73.844109   
2       17 2014-01-01 00:00:13 2014-01-01 00:18:35        -73.911644   
3       18 2014-01-01 00:00:16 2014-01-01 00:04:05        -73.954727   
4       19 2014-01-01 00:00:16 2014-01-01 00:08:32        -73.924431   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.771004         -73.888550         40.745453                1   
1        40.721107         -73.816330         40.714375                1   
2        40.767956         -73.946960         40.751373                1   
3        40.800072         -73.946892         40.813335                1   
4        40.743668         -73.934021         40.736256                1   

   trip_distance  
0           2.52  
1           1.66  
2           2.66  
3           1.02  
4           1.10

In [7]:
jan_green_one_rider.describe()

ride_id  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  640964.000000     640964.000000    640964.000000      640964.000000   
mean   403753.981295        -73.929510        40.768245         -73.930421   
std    232045.540272          0.041355         0.056710           0.049958   
min        14.000000        -74.331802        40.552734         -74.478737   
25%    202493.750000        -73.955269        40.721291         -73.963234   
50%    403116.500000        -73.940575        40.766838         -73.941170   
75%    604872.250000        -73.909353        40.810234         -73.903580   
max    803852.000000        -72.633492        41.183037         -72.633492   

       dropoff_latitude  passenger_count  trip_distance  
count     640964.000000         640964.0  640964.000000  
mean          40.764315              1.0       2.866513  
std            0.057734              0.0       2.828025  
min           40.441864              1.0       0.010000  
25%           40.724201              1.0       1.090000  
50%           40.763699              1.0       1.900000  
75%           40.807499              1.0       3.660000  
max           41.367435              1.0     115.610000

In [8]:
jan_green_one_rider.dtypes

ride_id                       int64
pickup_datetime      datetime64[ns]
dropoff_datetime     datetime64[ns]
pickup_longitude            float64
pickup_latitude             float64
dropoff_longitude           float64
dropoff_latitude            float64
passenger_count               int64
trip_distance               float64
dtype: object

In [9]:
print(round(jan_green_one_rider['trip_distance'].sum()), "total miles for ONE RIDER green taxi data in january 2014 from NYC")

1837331 total miles for ONE RIDER green taxi data in january 2014 from NYC


### Selecting the january green taxi data that was split with passenger count = 2

In [10]:
jan_green_two_rider = jan_green_two_rider = pd.read_sql("select * from jan_green_two_rider", db);

In [11]:
jan_green_two_rider.head()

ride_id     pickup_datetime    dropoff_datetime  pickup_longitude  \
0       24 2014-01-01 00:00:43 2014-01-01 00:07:44        -73.924568   
1       34 2014-01-01 00:01:05 2014-01-01 00:14:43        -73.929337   
2       36 2014-01-01 00:01:09 2014-01-01 00:03:59        -73.981842   
3       44 2014-01-01 00:01:36 2014-01-01 00:12:24        -73.953888   
4       67 2014-01-01 00:02:50 2014-01-01 00:05:51        -73.925217   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.861614         -73.942627         40.821888                2   
1        40.703571         -73.957573         40.729713                2   
2        40.666687         -73.985321         40.662525                2   
3        40.806480         -73.937546         40.797661                2   
4        40.761761         -73.917320         40.758137                2   

   trip_distance  
0           3.10  
1           3.30  
2           0.30  
3           1.21  
4           0.51

In [12]:
jan_green_two_rider.describe()

ride_id  pickup_longitude  pickup_latitude  dropoff_longitude  \
count   56464.000000      56464.000000     56464.000000       56464.000000   
mean   391011.861717        -73.930385        40.754977         -73.930964   
std    232746.447478          0.043352         0.055280           0.051819   
min        24.000000        -74.079437        40.573067         -74.484711   
25%    189143.750000        -73.957954        40.713748         -73.966669   
50%    397935.500000        -73.942066        40.747704         -73.942165   
75%    592494.250000        -73.904423        40.804214         -73.900877   
max    803851.000000        -73.247894        40.972569         -73.192268   

       dropoff_latitude  passenger_count  trip_distance  
count      56464.000000          56464.0   56464.000000  
mean          40.752764              2.0       3.011972  
std            0.056788              0.0       2.908302  
min           40.324463              2.0       0.010000  
25%           40.712924              2.0       1.130000  
50%           40.750050              2.0       2.050000  
75%           40.792474              2.0       3.900000  
max           41.233383              2.0      59.800000

In [13]:
jan_green_two_rider.dtypes

ride_id                       int64
pickup_datetime      datetime64[ns]
dropoff_datetime     datetime64[ns]
pickup_longitude            float64
pickup_latitude             float64
dropoff_longitude           float64
dropoff_latitude            float64
passenger_count               int64
trip_distance               float64
dtype: object

In [14]:
print(round(jan_green_two_rider['trip_distance'].sum()), "total miles for TWO RIDER green taxi data in january 2014 from NYC")

170068 total miles for TWO RIDER green taxi data in january 2014 from NYC


### Selecting the february green taxi data that was split with passenger count = 1

In [15]:
feb_green_one_rider = pd.read_sql("select * from feb_green_one_rider", db);

In [16]:
feb_green_one_rider.head()

ride_id     pickup_datetime    dropoff_datetime  pickup_longitude  \
0       36 2014-02-01 00:00:02 2014-02-01 00:08:29        -73.959465   
1       40 2014-02-01 00:00:08 2014-02-01 00:15:13        -73.990730   
2       42 2014-02-01 00:00:13 2014-02-01 00:06:56        -73.953674   
3       43 2014-02-01 00:00:24 2014-02-01 00:08:06        -73.944794   
4       44 2014-02-01 00:00:27 2014-02-01 00:19:34        -73.977577   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.716217         -73.965782         40.710552                1   
1        40.694611         -73.957352         40.714481                1   
2        40.790752         -73.959160         40.780113                1   
3        40.727329         -73.941681         40.708363                1   
4        40.678616         -73.988533         40.718513                1   

   trip_distance  
0           1.10  
1           3.50  
2           1.12  
3           1.40  
4           3.36

In [17]:
feb_green_one_rider.describe()

ride_id  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  8.151410e+05     815141.000000    815141.000000      815141.000000   
mean   5.045571e+05        -73.929019        40.767307         -73.929428   
std    2.904891e+05          0.041702         0.058114           0.049842   
min    3.600000e+01        -74.728561        40.294209         -75.989021   
25%    2.524790e+05        -73.955582        40.719196         -73.961983   
50%    5.046100e+05        -73.940231        40.765774         -73.939995   
75%    7.565040e+05        -73.906891        40.810978         -73.902313   
max    1.005242e+06        -73.054497        41.584133         -72.091438   

       dropoff_latitude  passenger_count  trip_distance  
count     815141.000000         815141.0  815141.000000  
mean          40.764263              1.0       2.846441  
std            0.059347              0.0       2.776918  
min           39.357197              1.0       0.010000  
25%           40.721554              1.0       1.100000  
50%           40.763062              1.0       1.910000  
75%           40.810101              1.0       3.630000  
max           41.585518              1.0     135.240000

In [18]:
feb_green_one_rider.dtypes

ride_id                       int64
pickup_datetime      datetime64[ns]
dropoff_datetime     datetime64[ns]
pickup_longitude            float64
pickup_latitude             float64
dropoff_longitude           float64
dropoff_latitude            float64
passenger_count               int64
trip_distance               float64
dtype: object

In [19]:
print(round(feb_green_one_rider['trip_distance'].sum()), "total miles for ONE RIDER green taxi data in february 2014 from NYC")

2320251 total miles for ONE RIDER green taxi data in february 2014 from NYC


### Selecting the february green taxi data that was split with passenger count = 2

In [20]:
feb_green_two_rider = pd.read_sql("select * from feb_green_two_rider", db);

In [21]:
feb_green_two_rider.head()

ride_id     pickup_datetime    dropoff_datetime  pickup_longitude  \
0       38 2014-02-01 00:00:08 2014-02-01 00:20:33        -73.961670   
1       39 2014-02-01 00:00:08 2014-02-01 00:28:26        -73.951134   
2       69 2014-02-01 00:00:59 2014-02-01 00:16:54        -73.960213   
3      126 2014-02-01 00:02:28 2014-02-01 00:30:48        -73.983704   
4      133 2014-02-01 00:02:36 2014-02-01 00:17:17        -73.935623   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.714012         -73.917099         40.746384                2   
1        40.748909         -73.980003         40.657005                2   
2        40.715313         -73.979050         40.678429                2   
3        40.676292         -73.983566         40.722137                2   
4        40.833282         -73.865150         40.920052                2   

   trip_distance  
0           3.88  
1          10.50  
2           4.35  
3           6.00  
4           8.60

In [22]:
feb_green_two_rider.describe()

ride_id  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  7.101700e+04      71017.000000     71017.000000       71017.000000   
mean   4.983607e+05        -73.930757        40.754266         -73.930883   
std    2.871839e+05          0.043510         0.056829           0.051334   
min    3.800000e+01        -74.177330        40.574635         -74.728554   
25%    2.548690e+05        -73.958336        40.711933         -73.965775   
50%    5.020950e+05        -73.942413        40.747322         -73.941162   
75%    7.474800e+05        -73.904327        40.804947         -73.900833   
max    1.005237e+06        -73.685745        40.983341         -73.306358   

       dropoff_latitude  passenger_count  trip_distance  
count      71017.000000          71017.0   71017.000000  
mean          40.752525              2.0       2.982883  
std            0.058684              0.0       2.849157  
min           39.346779              2.0       0.010000  
25%           40.710632              2.0       1.130000  
50%           40.749107              2.0       2.060000  
75%           40.796864              2.0       3.850000  
max           41.067734              2.0      76.380000

In [23]:
feb_green_two_rider.dtypes

ride_id                       int64
pickup_datetime      datetime64[ns]
dropoff_datetime     datetime64[ns]
pickup_longitude            float64
pickup_latitude             float64
dropoff_longitude           float64
dropoff_latitude            float64
passenger_count               int64
trip_distance               float64
dtype: object

In [24]:
print(round(feb_green_two_rider['trip_distance'].sum()), "total miles for TWO RIDER green taxi data in february 2014 from NYC")

211835 total miles for TWO RIDER green taxi data in february 2014 from NYC


### Selecting the march green taxi data that was split with passenger count = 1

In [25]:
mar_green_one_rider = pd.read_sql("select * from mar_green_one_rider", db);

In [26]:
mar_green_one_rider.head()

ride_id     pickup_datetime    dropoff_datetime  pickup_longitude  \
0       15 2014-03-01 00:00:00 2014-03-01 00:11:44        -73.807571   
1       34 2014-03-01 00:00:01 2014-03-01 00:04:27        -73.951355   
2       35 2014-03-01 00:00:03 2014-03-01 00:39:11        -73.958801   
3       36 2014-03-01 00:00:03 2014-03-01 00:14:32        -73.938881   
4       37 2014-03-01 00:00:03 2014-03-01 00:08:42        -73.941376   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.700371         -73.759422         40.704967                1   
1        40.809841         -73.937584         40.804348                1   
2        40.716785         -73.908257         40.698792                1   
3        40.681664         -73.956787         40.713566                1   
4        40.818493         -73.935242         40.796005                1   

   trip_distance  
0           3.14  
1           0.89  
2           7.05  
3           3.30  
4           2.38

In [27]:
mar_green_one_rider.describe()

ride_id  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  1.058404e+06      1.058404e+06     1.058404e+06       1.058404e+06   
mean   6.467195e+05     -7.393054e+01     4.076543e+01      -7.393090e+01   
std    3.729006e+05      4.134149e-02     5.825813e-02       4.988082e-02   
min    1.500000e+01     -7.436404e+01     4.031835e+01      -7.582375e+01   
25%    3.236348e+05     -7.395638e+01     4.071749e+01      -7.396370e+01   
50%    6.474105e+05     -7.394131e+01     4.076383e+01      -7.394112e+01   
75%    9.706468e+05     -7.391082e+01     4.081035e+01      -7.390433e+01   
max    1.293471e+06     -7.304129e+01     4.225634e+01      -7.241040e+01   

       dropoff_latitude  passenger_count  trip_distance  
count      1.058404e+06        1058404.0   1.058404e+06  
mean       4.076220e+01              1.0   2.923363e+00  
std        5.936801e-02              0.0   2.838458e+00  
min        3.958759e+01              1.0   1.000000e-02  
25%        4.071945e+01              1.0   1.100000e+00  
50%        4.076123e+01              1.0   1.980000e+00  
75%        4.080824e+01              1.0   3.740000e+00  
max        4.225480e+01              1.0   1.012600e+02

In [28]:
mar_green_one_rider.dtypes

ride_id                       int64
pickup_datetime      datetime64[ns]
dropoff_datetime     datetime64[ns]
pickup_longitude            float64
pickup_latitude             float64
dropoff_longitude           float64
dropoff_latitude            float64
passenger_count               int64
trip_distance               float64
dtype: object

In [29]:
print(round(mar_green_one_rider['trip_distance'].sum()), "total miles for ONE RIDER green taxi data in march 2014 from NYC")

3094100 total miles for ONE RIDER green taxi data in march 2014 from NYC


### Selecting the march green taxi data that was split with passenger count = 2

In [30]:
mar_green_two_rider = pd.read_sql("select * from mar_green_two_rider", db);

In [31]:
mar_green_two_rider.head()

ride_id     pickup_datetime    dropoff_datetime  pickup_longitude  \
0       41 2014-03-01 00:00:07 2014-03-01 00:11:19        -73.964630   
1       53 2014-03-01 00:00:16 2014-03-01 00:18:50        -73.917015   
2       62 2014-03-01 00:00:25 2014-03-01 00:05:02        -73.941803   
3       66 2014-03-01 00:00:28 2014-03-01 00:04:31        -73.962364   
4       92 2014-03-01 00:01:05 2014-03-01 00:04:24        -73.925407   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.712296         -73.947220         40.721889                2   
1        40.761211         -73.850166         40.725178                2   
2        40.798351         -73.941078         40.790031                2   
3        40.682636         -73.961952         40.689232                2   
4        40.761841         -73.935768         40.768932                2   

   trip_distance  
0           1.50  
1           7.17  
2           0.77  
3           0.70  
4           0.90

In [32]:
mar_green_two_rider.describe()

ride_id  pickup_longitude  pickup_latitude  dropoff_longitude  \
count  9.111900e+04      91119.000000     91119.000000       91119.000000   
mean   6.524365e+05        -73.932978        40.752387         -73.933037   
std    3.772053e+05          0.042651         0.056229           0.051104   
min    4.100000e+01        -74.438080        40.366768         -74.656998   
25%    3.275905e+05        -73.959290        40.709984         -73.968376   
50%    6.512600e+05        -73.944229        40.746983         -73.943207   
75%    9.737920e+05        -73.911263        40.803995         -73.904236   
max    1.293468e+06        -73.559784        41.034164         -73.040619   

       dropoff_latitude  passenger_count  trip_distance  
count      91119.000000          91119.0   91119.000000  
mean          40.750524              2.0       3.088483  
std            0.057751              0.0       2.969012  
min           40.354107              2.0       0.010000  
25%           40.708351              2.0       1.170000  
50%           40.748341              2.0       2.100000  
75%           40.792187              2.0       4.000000  
max           41.056427              2.0      87.660000

In [33]:
mar_green_two_rider.dtypes

ride_id                       int64
pickup_datetime      datetime64[ns]
dropoff_datetime     datetime64[ns]
pickup_longitude            float64
pickup_latitude             float64
dropoff_longitude           float64
dropoff_latitude            float64
passenger_count               int64
trip_distance               float64
dtype: object

In [34]:
print(round(mar_green_two_rider['trip_distance'].sum()), "total miles for TWO RIDER green taxi data in march 2014 from NYC")

281419 total miles for TWO RIDER green taxi data in march 2014 from NYC


# Algorithm with single and double riders

#### Function to get distance between coordinates

In [35]:
import geopy.distance

def get_distance(lat_origin, lon_origin, lat_end, lon_end):
    coords_1 = (lat_origin, lon_origin)
    coords_2 = (lat_end, lon_end)
    
    distance = geopy.distance.distance(coords_1, coords_2).mi
    #print("Distance: ", distance, " mi")
    return distance

get_distance(52.2296756, 21.0122287,52.406374, 16.9251681)

173.5818455248231

In [36]:
def time_difference(time_1, time_2):
    elapsedTime = time_2 - time_1
    #print("Elapsed time: ", elapsedTime)
    #print("Total seconds: ", elapsedTime.total_seconds())
    
    return (elapsedTime.total_seconds())

print("time difference from jan one riders : " + str(time_difference(jan_green_one_rider['pickup_datetime'][0], jan_green_one_rider['dropoff_datetime'][0])))
print("time difference from jan two riders : " + str(time_difference(jan_green_two_rider['pickup_datetime'][0], jan_green_two_rider['dropoff_datetime'][0])))

print("time difference from feb one riders : " + str(time_difference(feb_green_one_rider['pickup_datetime'][0], jan_green_one_rider['dropoff_datetime'][0])))
print("time difference from feb two riders : " + str(time_difference(feb_green_two_rider['pickup_datetime'][0], jan_green_two_rider['dropoff_datetime'][0])))

print("time difference from mar one riders : " + str(time_difference(mar_green_one_rider['pickup_datetime'][0], jan_green_one_rider['dropoff_datetime'][0])))
print("time difference from mar two riders : " + str(time_difference(mar_green_two_rider['pickup_datetime'][0], jan_green_two_rider['dropoff_datetime'][0])))

    

time difference from jan one riders : 501.0
time difference from jan two riders : 421.0
time difference from feb one riders : -2677894.0
time difference from feb two riders : -2677944.0
time difference from mar one riders : -5097092.0
time difference from mar two riders : -5097143.0


In [37]:
def add_speed_to_trip(trip):
    time_1 = trip['pickup_datetime']
    time_2 = trip['dropoff_datetime']

    distance = trip['trip_distance']
    time = time_difference(time_1, time_2)/3600
    speed = distance/time
    
    #print("Speed: ", speed, "mi/hr")
    return speed

In [38]:
jan_green_one_rider['speed'] = jan_green_one_rider.apply(add_speed_to_trip, axis = 1)
jan_green_two_rider['speed'] = jan_green_two_rider.apply(add_speed_to_trip, axis = 1)

feb_green_one_rider['speed'] = feb_green_one_rider.apply(add_speed_to_trip, axis = 1)
feb_green_two_rider['speed'] = feb_green_two_rider.apply(add_speed_to_trip, axis = 1)

mar_green_one_rider['speed'] = mar_green_one_rider.apply(add_speed_to_trip, axis = 1)
mar_green_two_rider['speed'] = mar_green_two_rider.apply(add_speed_to_trip, axis = 1)


In [39]:
def calcSequence(p1_lat, p1_long, p2_lat, p2_long, p3_lat, p3_long, p4_lat, p4_long, speed, trip1_pu_time, trip2_pu_time, 
                 trip1_do_time, trip2_do_time, delay):
  
    ## p1 - p2 - p3 - p4 / point1 - point2 - point3 - point4
    #within 5 minutes before -2 pickup, and withing 5minutes of after p3 and p4 dropoff
    
    #Keeps track of the total distance everytime distance is added
    totalDistance = 0
    
    # distance between o1 and o2  
    distance = get_distance(p1_lat,p1_long,p2_lat,p2_long)            
    # time it will take to go from o1 to o2
    time = speed/distance            
    # time is in hours so convert to minutes to seconds
    time_seconds = time * 60 * 60                       
    # delay in seconds
    delay_seconds = delay * 60            
    # trip 2 pickupd time range edge
                   
    # [trip2_pickup_edge, trip2_pickup]  
    # [12:55 pm, 1:00 pm]
    # trip1_pickup + time seconds = some time that needs to fall in this range            
    if (trip1_pu_time + time_seconds >= (trip2_pu_time - delay_seconds)) or (trip1_pu_time + time_seconds <= trip2_pu_time):
        # Get the trip1 drop off edge
        # p2 - p3
        totalDistance += distance
        distance = get_distance(p2_lat,p2_long,p3_lat,p3_long)
       
        # time it will take to go from o1 to o2
        time = speed/distance            
        # time is in hours so convert to minutes to seconds
        time_seconds = time * 60 * 60  
        if (trip2_pickup + time_seconds >= (trip1_do_time)) or (trip2_pu_time + time_seconds <= trip1_do_time + delay_seconds):
            totalDistance += distance
            distance = get_distance(p3_lat,p3_long,p4_lat,p4_long)   
            # time it will take to go from o1 to o2
            time = speed/distance            
            # time is in hours so convert to minutes to seconds
            time_seconds = time * 60 * 60  
            if (trip1_do_time + time_seconds >= (trip2_do_time)) or (trip1_do_time + time_seconds <= trip2_do_time  + delay_seconds):
                totalDistance += distance
                return totalDistance
    return -1
        


    # get time range permissible for each origin and dropoff
    # start at o1 check if taxi can make it to o2 within delay

    # i.e. if 5 minute delay then make it within 5 min before o2
    # then have to make it within 5 min after d1 and d2 drop off

    # calculate savings if trip is permissible
    # savings = [distance(trip1) + distance(trip2)] - distance(combined trip1 & trip2)

    # o1 - o2 - d2 - d1

    # o2 - o1 - d1 - d2

    # o2 - o1 - d2 - d1
    


def prototype_algorithm(single_trips, double_trips, delay):
    # columns- 'ride_id', 'pickup_datetime', 'dropoff_datetime', 'pickup_longitude',
    #          'pickup_latitude', 'dropoff_longitude', 'dropoff_latitude',
    #          'passenger_count', 'trip_distance', 'speed'
    combined_trips = {}

    for index, trip1 in single_trips.iterrows():  

        for index, trip2 in double_trips.iterrows():
            # average speed of two trips
            speed3 = (trip1['speed'] + trip2['speed'])/2
                        
            # possible sequences:
            
            # o1 - o2 - d1 - d2
            
            # o1 - o2 - d2 - d1
            
            # o2 - o1 - d1 - d2
            
            # o2 - o1 - d2 - d1
            o1_lon = trip1['pickup_longitude']
            o1_lat = trip1['pickup_latitude']
            o2_lon = trip2['pickup_longitude']
            o2_lat = trip2['pickup_latitude']
            
            d1_lon = trip1['pickup_longitude']
            d1_lat = trip1['pickup_latitude']
            d2_lon = trip2['pickup_longitude']
            d2_lat = trip2['pickup_latitude']
            
            
            pu1 = trip1['pickup_datetime'].total_seconds()  
            pu2 = trip2['pickup_datetime'].total_seconds() 
            
            
            do1 = trip1['dropoff_datetime'].total_seconds()
            do2 = trip2['dropoff_datetime'].total_seconds() 
            
            
            # o1 - o2 - d1 - d2
            seq1 = calcSequence(o1_lat, o1_long, o2_lat, o2_long, d1_lat, d1_long, d2_lat, d2_long, speed3, pu1, pu2, du1, du2, delay)
            
            # o1 - o2 - d2 - d1
            seq2 = calcSequence(o1_lat, o1_long, o2_lat, o2_long, d2_lat, d2_long, d1_lat, d1_long, speed3, pu1, pu2, du1, du2, delay)
            
            # o2 - o1 - d2 - d1
            seq3 = calcSequence( o2_lat, o2_long, o1_lat, o1_long, d2_lat, d2_long, d1_lat, d1_long, speed3, pu1, pu2, du1, du2, delay)
            
            # o2 - o1 - d1 - d2
            seq4 = calcSequence( o2_lat, o2_long, o1_lat, o1_long, d1_lat, d1_long, d2_lat, d2_long, speed3, pu1, pu2, du1, du2, delay)
            
            totalDistance = trip1['trip_distance'] + trip2['trip_distance']
            
            
            if seq1 < seq2 and seq1 < seq3 and seq1 < seq4 and seq1 != -1:
                combined_trips[(trip1['ride_id'], trip2['ride_id'])] = totalDistance - seq1
            elif seq2 < seq1 and seq2 < seq3 and seq2 < seq4 and seq2 != -1:
                combined_trips[(trip1['ride_id'], trip2['ride_id'])] = totalDistance - seq2   
            elif seq3 < seq1 and seq3 < seq4 and seq3 < seq2 and seq3 != -1:
                combined_trips[(trip1['ride_id'], trip2['ride_id'])] = totalDistance - seq3 
            elif seq4 < seq1 and seq4 < seq2 and seq4 < seq3 and seq4 != -1:
                combined_trips[(trip1['ride_id'], trip2['ride_id'])] = totalDistance - seq4     
            
    return combined_trips


##### Perhaps Limit the Longitude and Latitude ranges to make algorithm faster
These values are chosen based on the 25% to 50% quartile ranges for the latitude and longitude to get 25% of the data isntead of 100%

-73.959290	< pu_lon < -73.944229 

40.709984	< pu_lat < 40.746983	

-73.968376	< do_lon < -73.943207	

40.708351	< do_lat < 40.748341	


In [40]:
jan_one = jan_green_one_rider
jan_two = jan_green_two_rider

In [41]:
jan_one.shape

(640964, 10)

In [42]:
jan_two.shape

(56464, 10)

In [43]:
jan_one = jan_one[(jan_one['pickup_latitude'] > 40.709984) & (jan_one['pickup_latitude'] < 40.746983) &
                  (jan_one['pickup_longitude'] > -73.959290) & (jan_one['pickup_longitude'] < -73.944229) &
                  (jan_one['dropoff_latitude'] > 40.708351) & (jan_one['dropoff_latitude'] < 40.748341) &
                  (jan_one['dropoff_longitude'] > -73.968376) & (jan_one['dropoff_longitude'] < -73.943207)
                 ]

In [44]:
jan_two = jan_two[(jan_two['pickup_latitude'] > 40.709984) & (jan_two['pickup_latitude'] < 40.746983) &
                  (jan_two['pickup_longitude'] > -73.959290) & (jan_two['pickup_longitude'] < -73.944229) &
                  (jan_two['dropoff_latitude'] > 40.708351) & (jan_two['dropoff_latitude'] < 40.748341) &
                  (jan_two['dropoff_longitude'] > -73.968376) & (jan_two['dropoff_longitude'] < -73.943207)
                 ]

In [45]:
jan_one.shape

(13676, 10)

In [46]:
jan_two.shape

(1813, 10)

In [47]:
jan_one = jan_one.sort_values(by = ['pickup_datetime', 'pickup_latitude', 'pickup_longitude']).reset_index().drop(columns=['index'])
jan_one.head()

ride_id     pickup_datetime    dropoff_datetime  pickup_longitude  \
0       38 2014-01-01 00:01:16 2014-01-01 00:07:05        -73.949501   
1       59 2014-01-01 00:02:10 2014-01-01 00:04:54        -73.951622   
2       60 2014-01-01 00:02:11 2014-01-01 00:06:53        -73.957886   
3       64 2014-01-01 00:02:39 2014-01-01 00:06:04        -73.952599   
4       76 2014-01-01 00:03:16 2014-01-01 00:13:14        -73.951637   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.714016         -73.947380         40.710068                1   
1        40.714329         -73.950470         40.711063                1   
2        40.717773         -73.957207         40.727459                1   
3        40.726860         -73.945587         40.725471                1   
4        40.713833         -73.957687         40.732048                1   

   trip_distance      speed  
0           0.88   9.077364  
1           0.47  10.317073  
2           0.89  11.361702  
3           0.90  15.804878  
4           1.80  10.836120

In [48]:
jan_two = jan_two.sort_values(by = ['pickup_datetime', 'pickup_latitude', 'pickup_longitude']).reset_index().drop(columns=['index'])
jan_two.head()

ride_id     pickup_datetime    dropoff_datetime  pickup_longitude  \
0       86 2014-01-01 00:03:43 2014-01-01 00:11:39        -73.957870   
1      568 2014-01-01 00:16:16 2014-01-01 00:24:40        -73.957764   
2      954 2014-01-01 00:24:27 2014-01-01 00:32:28        -73.950150   
3     1117 2014-01-01 00:28:14 2014-01-01 00:32:58        -73.945168   
4     1373 2014-01-01 00:33:45 2014-01-01 00:39:54        -73.955292   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.721825         -73.943550         40.708923                2   
1        40.722015         -73.946861         40.708637                2   
2        40.718609         -73.955276         40.736778                2   
3        40.719082         -73.954559         40.732052                2   
4        40.736786         -73.956200         40.719257                2   

   trip_distance      speed  
0           1.47  11.117647  
1           1.60  11.428571  
2           1.67  12.498960  
3           1.10  13.943662  
4           1.58  15.414634

## Algorithm

### Select one rider

In [49]:
# Algorithm
# select one ride
ride_one = jan_one.loc[[0]]
ride_one

ride_id     pickup_datetime    dropoff_datetime  pickup_longitude  \
0       38 2014-01-01 00:01:16 2014-01-01 00:07:05        -73.949501   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.714016          -73.94738         40.710068                1   

   trip_distance     speed  
0           0.88  9.077364

### Select two rider

In [50]:
ride_two = jan_two.loc[[0]]
ride_two

ride_id     pickup_datetime    dropoff_datetime  pickup_longitude  \
0       86 2014-01-01 00:03:43 2014-01-01 00:11:39         -73.95787   

   pickup_latitude  dropoff_longitude  dropoff_latitude  passenger_count  \
0        40.721825          -73.94355         40.708923                2   

   trip_distance      speed  
0           1.47  11.117647

### Check if combinable with delay 5 minutes

In [51]:
delay = 5
print("Delay:", delay, "minutes")

Delay: 5 minutes


#### 1. get speed of combined trips

In [52]:
speed = (ride_one['speed'][0] + ride_two['speed'][0])/2
print(speed, "mi/hr")

10.097505477835835 mi/hr


#### 2a. get distance between  pickup points

In [53]:
pu_distance = get_distance(ride_one['pickup_latitude'][0], ride_one['pickup_longitude'][0],
                           ride_two['pickup_latitude'][0], ride_two['pickup_longitude'][0])
print(pu_distance, "mi")

0.6952721673692104 mi


#### 2b. get distance between dropoff points

In [54]:
do_distance = get_distance(ride_one['dropoff_latitude'][0], ride_one['dropoff_longitude'][0],
                           ride_two['dropoff_latitude'][0], ride_two['dropoff_longitude'][0])
print(do_distance, "mi")

0.2160517717605075 mi


#### 3a. get time it will take to travel distance between PU points

In [55]:
pu_travel_time = (pu_distance/speed) * 60 # in hours, change to minute by multiplying 60
print(pu_travel_time, "minutes")

4.131350077870277 minutes


#### 3b. get time it will take to travel distance between DO points

In [56]:
do_travel_time = (do_distance/speed) * 60 # in hours, change to minute by multiplying 60
print(do_travel_time, "minutes")

1.28379294609789 minutes


#### 4a. compare PU times to delay + time from 3a to see if trip is likely valid

In [57]:
from datetime import datetime as dt
import pandas as pandas

In [58]:
# get pickup times for both rides

ride_one_pu_time = (pandas.to_datetime(ride_one['pickup_datetime'][0]).second +
                    pandas.to_datetime(ride_one['pickup_datetime'][0]).minute*60 +
                    pandas.to_datetime(ride_one['pickup_datetime'][0]).hour*60*60)

ride_two_pu_time = (pandas.to_datetime(ride_two['pickup_datetime'][0]).second +
                    pandas.to_datetime(ride_two['pickup_datetime'][0]).minute*60 +
                    pandas.to_datetime(ride_two['pickup_datetime'][0]).hour*60*60)

distance = -1
trip_sequence = []
# combo: o1 -> o2
# have to make it by 5 minutes before pickup 2
if (ride_one_pu_time + pu_travel_time) < (ride_two_pu_time - delay): 
    distance = pu_distance
    combo = (ride_one['ride_id'][0], ride_two['ride_id'][0])
# combo: o2 -> o1
# have to make it by 5 minutes before pickup 1
if (ride_two_pu_time + pu_travel_time) < (ride_one_pu_time - delay): 
    distance = pu_distance
    combo = (ride_two['ride_id'][0], ride_one['ride_id'][0])
    
if distance == -1:
    print("FAIL")
else:
    print("Ride 1 id: ", ride_one['ride_id'][0])
    print("Ride 2 id: ", ride_two['ride_id'][0])
    print()
    print("Trip combinable for pickup segment with distance", distance)
    print("Combinable sequence is: ", combo)

Ride 1 id:  38
Ride 2 id:  86

Trip combinable for pickup segment with distance 0.6952721673692104
Combinable sequence is:  (38, 86)


#### 4b.  compare DO times to delay + time from 3b to see if trip is likely valid

In [59]:
do_time_difference = time_difference(ride_one['dropoff_datetime'][0], ride_two['dropoff_datetime'][0])
print(do_time_difference, "seconds")

274.0 seconds


In [60]:
# get dropoff times for both rides

ride_one_do_time = (pandas.to_datetime(ride_one['dropoff_datetime'][0]).second +
                    pandas.to_datetime(ride_one['dropoff_datetime'][0]).minute*60 +
                    pandas.to_datetime(ride_one['dropoff_datetime'][0]).hour*60*60)

ride_two_do_time = (pandas.to_datetime(ride_two['dropoff_datetime'][0]).second +
                    pandas.to_datetime(ride_two['dropoff_datetime'][0]).minute*60 +
                    pandas.to_datetime(ride_two['dropoff_datetime'][0]).hour*60*60)

distance = -1
trip_sequence = []
# combo: d1 -> d2
# have to make it by 5 minutes before pickup 2
if (ride_one_do_time + do_travel_time) < (ride_two_do_time + delay): 
    distance = do_distance
    combo = (ride_one['ride_id'][0], ride_two['ride_id'][0])

# combo: d2 -> d1
# have to make it by 5 minutes before pickup 1
if (ride_two_do_time + do_travel_time) < (ride_one_do_time + delay): 
    distance = do_distance
    combo = (ride_two['ride_id'][0], ride_one['ride_id'][0])
    
if distance == -1:
    print("FAIL")
else:
    print("Ride 1 id: ", ride_one['ride_id'][0])
    print("Ride 2 id: ", ride_two['ride_id'][0])
    print()
    print("Trip combinable for pickup segment with distance", distance)
    print("Combinable sequence is: ", combo)

Ride 1 id:  38
Ride 2 id:  86

Trip combinable for pickup segment with distance 0.2160517717605075
Combinable sequence is:  (38, 86)


##### Analyzing results

For combining trip in the pickup segment the order of trips (ride_one id, ride_two id) = (38, 86) also can be said to be (o1, o2) is a valid order for delay given 

For combining tip in the dropoff segment the order of trips (ride_one id, ride_two id) = (38, 86) also can be said to be (d1, d2) is a valid order for delay given 

##### Hypothesis:
If the pickup segment and dropoff segment of ridesharing is valid then the center part of the trip that varies in distance should be valid

If a valid solution is found then there is an optimal combination for the rideshare.

#### 5.  Checking if sequence will result in saved distance for combining ride

In [61]:
# Sequence is o1, o2, d1, d2

# get distance between o2, d1
distance = get_distance(ride_two['pickup_latitude'][0], ride_two['pickup_longitude'][0],
                        ride_one['dropoff_latitude'][0], ride_one['dropoff_longitude'][0])
print("Distance between o2, d1: ", distance, "mi")

# get the time it will take to travel in the center part of the trip from o2 to d1
center_travel_time = distance/speed

# check if will make it within delay
if (ride_two_pu_time + center_travel_time) < (ride_one_do_time + delay):
    rideshare_distance = pu_distance + distance + do_distance

    print("Rideshare total trip distance: ", rideshare_distance, "mi")


Distance between o2, d1:  0.9805547868624445 mi
Rideshare total trip distance:  1.8918787259921623 mi


#### 6.  Seeing if there was a saving in distance

In [62]:
original_distance = ride_one['trip_distance'][0] + ride_two['trip_distance'][0]
print("Original total distance of the individual trips: ", original_distance, "mi")

Original total distance of the individual trips:  2.35 mi


##### Results:
The combined rideshare trip was about 1.89 mi

The total of the individual rides is 2.35 mi


#### 7.  Getting distance savings

In [63]:
distance_savings = original_distance - rideshare_distance
passenger_count = ride_one['passenger_count'][0] + ride_two['passenger_count'][0]

print("The distance saved by combining ride 1 with id", ride_one['ride_id'][0], 
      "and combining ride 2 with id", ride_two['ride_id'][0], "is", 
      distance_savings, "mi", "\nwithin a delay of", delay, "minutes and a total of",
      passenger_count, "passengers"
     )

The distance saved by combining ride 1 with id 38 and combining ride 2 with id 86 is 0.45812127400783775 mi 
within a delay of 5 minutes and a total of 3 passengers


#### 8.  Storing the ride combos and information such as distance, savings, and passenger count in a dictionary

In [68]:
# Combined rides list is "ride one id, ride two id, distance of combined trip, saved milage, percent milage saved, passenger count for combined trip"
combined_rides_dict = {}

combined_rides_dict[ride_one['ride_id'][0]] = [ride_two['ride_id'][0], rideshare_distance, distance_savings, distance_savings/original_distance*100, passenger_count]
combined_rides_dict[ride_two['ride_id'][0]] = [ride_one['ride_id'][0], rideshare_distance, distance_savings, distance_savings/original_distance*100, passenger_count]

print(combined_rides_dict)

{38: [86, 1.8918787259921623, 0.45812127400783775, 19.49452229820586, 3], 86: [38, 1.8918787259921623, 0.45812127400783775, 19.49452229820586, 3]}


##### Function to check if ride has already been combined

In [65]:
def check_if_combined(ride_id):
    return(ride_id in combined_rides_dict)

In [66]:
# Ride id 38 already combined with a trip from before so function should return true
print(check_if_combined(38))

True


In [67]:
# Ride id 3 has not been combined with a trip so function should return false
print(check_if_combined(3))

False
